sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroEncoder.ts

```diff
@@ -0,0 +1,217 @@
+ * pack schemas into its payloads.
+ */
+export class SchemaRegistryAvroEncoder<MessageT = MessageWithMetadata> {
+  /**
+   * Creates a new encoder.
+   *
+   * @param client - Schema Registry where schemas are registered and obtained.
+   *                 Usually this is a SchemaRegistryClient instance.
+   */
+  constructor(client: SchemaRegistry, options?: SchemaRegistryAvroEncoderOptions<MessageT>) {
+    this.registry = client;
+    this.schemaGroup = options?.groupName;
+    this.autoRegisterSchemas = options?.autoRegisterSchemas ?? false;
+    this.messageAdapter = options?.messageAdapter;
+  }
+
+  private readonly schemaGroup?: string;
+  private readonly registry: SchemaRegistry;
+  private readonly autoRegisterSchemas: boolean;
+  private readonly messageAdapter?: MessageAdapter<MessageT>;
+
+  // REVIEW: signature.
+  //
+  // - Should we wrap all errors thrown by avsc to avoid having our exception //
+  //   contract being tied to its implementation details?
+  /**
+   * encodes the value parameter according to the input schema and creates a message
+   * with the encoded data.
+   *
+   * @param value - The value to encodeMessageData.
+   * @param schema - The Avro schema to use.
+   * @param options - Encoding options.
+   * @returns A new message with the encoded value. The structure of message is
+   * constrolled by the message factory option.
+   */
+  async encodeMessageData(value: unknown, schema: string): Promise<MessageT> {
+    const entry = await this.getSchemaByDefinition(schema);
+    const buffer = entry.type.toBuffer(value);
+    const payload = new Uint8Array(
+      buffer.buffer,
+      buffer.byteOffset,
+      buffer.byteLength / Uint8Array.BYTES_PER_ELEMENT
+    );
+    const contentType = `${avroMimeType}+${entry.id}`;
+    return this.messageAdapter
+      ? this.messageAdapter.produceMessage({
+          contentType,
+          body: payload,
+        })
+      : /**
+         * If no message consumer was provided, then a MessageWithMetadata will be
+         * returned. This should work because the MessageT type parameter defaults
+         * to MessageWithMetadata.
+         */
+        ({
+          body: payload,
+          contentType: contentType,
+        } as unknown as MessageT);
+  }
+
+  /**
+   * Decodes the payload of the message using the schema ID in the content type
+   * field if no schema was provided.
+   *
+   * @param message - The message with the payload to be decoded.
+   * @param options - Decoding options.
+   * @returns The decoded value.
+   */
+  async decodeMessageData(
+    message: MessageT,
+    options: DecodeMessageDataOptions = {}
+  ): Promise<unknown> {
+    const { schema: readerSchema } = options;
+    const { body, contentType } = getPayloadAndContent(message, this.messageAdapter);
+    const buffer = Buffer.from(body);
+    if (readerSchema) {
+      return this.getAvroTypeForSchema(readerSchema).fromBuffer(buffer);
```

Comment:
I'm a bit confused why we don't need to get the schemaId if the user passed in a readerSchema. I'd expect that we would still need the writer schema because we need to know how the data was encoded. If the reader and writer schemas differ, would this code work?

Reply:
Great catch, I did not realize the writer schema is needed but of course it is to compute where to slice in the buffer. I fixed this and added a couple unit tests in https://github.com/Azure/azure-sdk-for-js/pull/19842/commits/5e2a21d56ed9530fe7c39b7fc1a288ec86d6115f.